### PR TITLE
add diagnostic information when incosistency is found

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Add diagnostic information when incosistency is found between primary index and documents cf. Also validate document existence and lock document keys properly when isRestore=true and we are in maintainer mode.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: -
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBKey.cpp [hunk 1]
```diff
@@ -29,6 +29,8 @@
 #include "RocksDBEngine/RocksDBFormat.h"
 #include "RocksDBEngine/RocksDBTypes.h"
 
+#include <iostream>
```
Comment (neunhoef) [inline arangod/RocksDBEngine/RocksDBKey.cpp:32]: Needed?

File: arangod/RocksDBEngine/RocksDBKey.h [hunk 2]
```diff
@@ -31,6 +31,7 @@
 #include "VocBase/voc-types.h"
 #include "Zkd/ZkdHelper.h"
 
+#include <iosfwd>
```
Comment (neunhoef) [inline arangod/RocksDBEngine/RocksDBKey.h:34]: Needed?

**Review by neunhoef — APPROVED** [review]
LGTM. See two comments, two include statements seem to be superfluous.

File: arangod/RocksDBEngine/RocksDBKey.h (hunk 2 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBKey.h:34]: I think so, as the file is implementing `operator<<` to append to an `std::ostream`.
It already did so before, but the required headers were missing.

File: arangod/RocksDBEngine/RocksDBKey.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBKey.cpp:32]: I think so, as the file is implementing `operator<<` to append to an `std::ostream`.
It already did so before, but the required headers were missing.

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 3]
```diff
@@ -1405,11 +1417,27 @@ Result RocksDBCollection::insertDocument(transaction::Methods* trx,
...
+  rocksdb::Status s;
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  // use Put() instead of PutUntracked()
+  if (options.isRestore) {
+    s = mthds->Put(RocksDBColumnFamilyManager::get(
+                       RocksDBColumnFamilyManager::Family::Documents),
+                   key.ref(),
+                   rocksdb::Slice(doc.startAs<char>(),
+                                  static_cast<size_t>(doc.byteSize())),
+                   /*assume_tracked*/ false);
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1429]: `Put` will only check for conflicts with concurrent operations, but would still silently overwrite any already existing key. We should instead use a `GetForUpdate` to check for an existing key and lock it. Then we can even stick with `PutUntracked`.

**Review by mpoeter — CHANGES_REQUESTED** [review]

File: arangod/RocksDBEngine/RocksDBCollection.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBCollection.cpp]: Addressed.

**Review by mpoeter — APPROVED** [review]

**Pull request merged**